File: projects/state-machine/python/src/state_machine.py
"""
Core State Machine implementation.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Callable, Dict, List, Optional, Set, Tuple, Union

from .action import Action, EntryAction, ExitAction, FunctionAction, TransitionAction
from .error import (
    ActionError,
    ConfigurationError,
    GuardRejectedError,
    InvalidEventError,
    InvalidStateError,
    TransitionError,
)
from .event import Event
from .guard import Guard
from .history import HistoryEntry, HistoryManager, HistoryState
from .state import State
from .transition import Transition, TransitionBuilder


class StateMachine:
    """
    A finite state machine implementation.

    Supports:
    - State definitions with entry/exit actions
    - Event-driven transitions
    - Guard conditions
    - Transition actions
    - History recording
    - Callbacks for state changes and failures

    Example:
        >>> sm = StateMachine(State("Off"))
        >>> sm.add_transition(Transition(
        ...     from_state=State("Off"),
        ...     to_state=State("On"),
        ...     event=Event("TurnOn")
        ... ))
        >>> sm.process_event(Event("TurnOn"))
        >>> sm.current_state
        State('On')
    """

    def __init__(
        self,
        initial_state: State,
        enable_history: bool = True,
        history_capacity: int = 1000,
    ) -> None:
        """
        Initialize the state machine.

        Args:
            initial_state: The initial state
            enable_history: Whether to record transition history
            history_capacity: Maximum history entries to keep
        """
        self._initial_state: State = initial_state
        self._current_state: State = initial_state
        self._transitions: List[Transition] = []
        self._states: Set[State] = {initial_state}
        self._events: Set[Event] = set()

        # History
        self._enable_history = enable_history
        self._history: Optional[HistoryManager] = (
            HistoryManager(history_capacity) if enable_history else None
        )
        self._history_state = HistoryState()

        # Callbacks
        self._on_state_change: Optional[Callable] = None
        self._on_transition_failed: Optional[Callable] = None
        self._on_event_processed: Optional[Callable] = None

        # Context
        self._context: Dict[str, Any] = {}

    @property
    def current_state(self) -> State:
        """Get the current state."""
        return self._current_state

    @property
    def states(self) -> Set[State]:
        """Get all defined states."""
        return self._states.copy()

    @property
    def events(self) -> Set[Event]:
        """Get all defined events."""
        return self._events.copy()

    @property
    def transitions(self) -> List[Transition]:
        """Get all defined transitions."""
        return self._transitions.copy()

    @property
    def history(self) -> Optional[HistoryManager]:
        """Get the history manager."""
        return self._history

    @property
    def context(self) -> Dict[str, Any]:
        """Get the current context."""
        return self._context

    @context.setter
    def context(self, value: Dict[str, Any]) -> None:
        """Set the context."""
        self._context = value

    def add_transition(self, transition: Transition) -> None:
        """
        Add a transition to the state machine.

        Args:
            transition: The transition to add

        Raises:
            ConfigurationError: If the transition is invalid
        """
        # Validate states exist
        self._states.add(transition.from_state)
        self._states.add(transition.to_state)
        self._events.add(transition.event)

        # Check for duplicate transitions
        for existing in self._transitions:
            if existing.matches(transition.from_state, transition.event):
                if existing.to_state == transition.to_state:
                    raise ConfigurationError(
                        f"Duplicate transition: {transition.from_state} -> "
                        f"{transition.to_state} on {transition.event}"
                    )

        self._transitions.append(transition)

    def add_transitions(self, transitions: List[Transition]) -> None:
        """
        Add multiple transitions.

        Args:
            transitions: List of transitions to add
        """
        for transition in transitions:
            self.add_transition(transition)

    def on_state_change(self, callback: Callable) -> None:
        """
        Set callback for state changes.

        Args:
            callback: Function(from_state, to_state, event)
        """
        self._on_state_change = callback

    def on_transition_failed(self, callback: Callable) -> None:
        """
        Set callback for failed transitions.

        Args:
            callback: Function(from_state, event, error)
        """
        self._on_transition_failed = callback

    def on_event_processed(self, callback: Callable) -> None:
        """
        Set callback for event processing.

        Args:
            callback: Function(event, success)
        """
        self._on_event_processed = callback

    def process_event(
        self,
        event: Event,
        context: Optional[Dict[str, Any]] = None,
    ) -> bool:
        """
        Process an event, triggering a transition if applicable.

        Args:
            event: The event to process
            context: Optional context data (merged with instance context)

        Returns:
            True if a transition occurred, False otherwise

        Raises:
            InvalidEventError: If no transition is defined for the event
        """
        # Merge contexts
        merged_context = {**self._context}
        if context:
            merged_context.update(context)

        # Find matching transitions
        matching_transitions = [
            t for t in self._transitions
            if t.matches(self._current_state, event)
        ]

        if not matching_transitions:
            error = InvalidEventError(event, self._current_state)
            if self._on_transition_failed:
                self._on_transition_failed(self._current_state, event, error)
            if self._enable_history and self._history:
                self._history.push(HistoryEntry(
                    timestamp=datetime.now(),
                    from_state=self._current_state,
                    to_state=self._current_state,
                    event=event,
                    success=False,
                    context=merged_context,
                    error=error,
                ))
            return False

        # Try each matching transition
        for transition in matching_transitions:
            if transition.can_transition(merged_context):
                return self._execute_transition(transition, merged_context)

        # All guards rejected
        error = GuardRejectedError(
            self._current_state,
            matching_transitions[0].to_state,
            event,
        )
        if self._on_transition_failed:
            self._on_transition_failed(self._current_state, event, error)
        if self._enable_history and self._history:
            self._history.push(HistoryEntry(
                timestamp=datetime.now(),
                from_state=self._current_state,
                to_state=matching_transitions[0].to_state,
                event=event,
                success=False,
                context=merged_context,
                error=error,
            ))
        return False

    def _execute_transition(
        self,
        transition: Transition,
        context: Dict[str, Any],
    ) -> bool:
        """
        Execute a transition.

        Args:
            transition: The transition to execute
            context: The merged context

        Returns:
            True if successful
        """
        from_state = self._current_state
        to_state = transition.to_state
        event = transition.event

        try:
            # Execute exit action on current state
            from_state.on_exit(context)

            # Execute transition action
            transition.execute(context)

            # Update state
            self._current_state = to_state

            # Execute entry action on new state
            to_state.on_enter(context)

            # Record history
            if self._enable_history and self._history:
                entry = HistoryEntry(
                    timestamp=datetime.now(),
                    from_state=from_state,
                    to_state=to_state,
                    event=event,
                    success=True,
                    context=context,
                )
                self._history.push(entry)

            # Notify callback
            if self._on_state_change:
                self._on_state_change(from_state, to_state, event)

            if self._on_event_processed:
                self._on_event_processed(event, True)

            return True

        except Exception as e:
            error = ActionError(
                f"Action failed during transition: {e}",
                original_error=e,
            )
            if self._on_transition_failed:
                self._on_transition_failed(from_state, event, error)
            if self._enable_history and self._history:
                entry = HistoryEntry(
                    timestamp=datetime.now(),
                    from_state=from_state,
                    to_state=to_state,
                    event=event,
                    success=False,
                    context=context,
                    error=error,
                )
                self._history.push(entry)
            return False

    def can_process_event(self, event: Event) -> bool:
        """
        Check if an event can be processed.

        Args:
            event: The event to check

        Returns:
            True if a transition exists and guard passes
        """
        for transition in self._transitions:
            if transition.matches(self._current_state, event):
                if transition.can_transition(self._context):
                    return True
        return False

    def possible_events(self) -> List[Event]:
        """
        Get all possible events from the current state.

        Returns:
            List of events that have defined transitions
        """
        return [
            t.event for t in self._transitions
            if t.from_state == self._current_state
        ]

    def possible_transitions(self) -> List[Transition]:
        """
        Get all possible transitions from the current state.

        Returns:
            List of transitions from the current state
        """
        return [
            t for t in self._transitions
            if t.from_state == self._current_state
        ]

    def set_state(self, state: State) -> None:
        """
        Directly set the current state (without transition).

        Args:
            state: The state to set

        Raises:
            InvalidStateError: If the state is not defined
        """
        if state not in self._states:
            raise InvalidStateError(state)
        self._current_state = state

    def reset(self, state: Optional[State] = None) -> None:
        """
        Reset the state machine to initial or specified state.

        Args:
            state: Optional state to reset to
        """
        if state:
            self._current_state = state
        else:
            self._current_state = self._initial_state
        if self._history:
            self._history.clear()

    def format_transitions(self) -> str:
        """
        Format all transitions as a string.

        Returns:
            Formatted string of all transitions
        """
        if not self._transitions:
            return "No transitions defined"
        return "\n".join(str(t) for t in self._transitions)

    def format_graph(self) -> str:
        """
        Format the state machine as a DOT graph.

        Returns:
            DOT format graph string
        """
        lines = ["digraph StateMachine {"]
        lines.append('  rankdir=LR;')
        lines.append('  node [shape=circle];')

        # Mark initial state
        lines.append(f'  "" [shape=point];')
        lines.append(f'  "" -> "{self._current_state}";')

        # Add transitions
        for t in self._transitions:
            label = str(t.event)
            if t.description:
                label = t.description
            lines.append(f'  "{t.from_state}" -> "{t.to_state}" [label="{label}"];')

        lines.append("}")
        return "\n".join(lines)

    def __repr__(self) -> str:
        return (
            f"StateMachine(state={self._current_state}, "
            f"transitions={len(self._transitions)}, "
            f"states={len(self._states)})"
        )


def create_state_machine(
    initial_state: State,
    transitions: Optional[List[Transition]] = None,
    enable_history: bool = True,
) -> StateMachine:
    """
    Factory function to create a state machine.

    Args:
        initial_state: The initial state
        transitions: Optional list of transitions
        enable_history: Whether to enable history

    Returns:
        A configured StateMachine instance

    Example:
        >>> sm = create_state_machine(
        ...     State("Off"),
        ...     transitions=[
        ...         Transition(State("Off"), State("On"), Event("TurnOn")),
        ...         Transition(State("On"), State("Off"), Event("TurnOff")),
        ...     ]
        ... )
    """
    sm = StateMachine(initial_state, enable_history)
    if transitions:
        sm.add_transitions(transitions)
    return sm

File: projects/state-machine/src/state_machine.rs
use crate::error::StateMachineError;
use crate::history::{HistoryEntry, HistoryManager};
use crate::transition::Transition;
use crate::{Event, Result, State};
use log::{debug, info, warn};
use std::collections::HashMap;
use std::fmt::Debug;
use std::time::Instant;

/// Type alias for state change callback
type StateChangeCallback<S, E> = Box<dyn Fn(&S, &S, &E) + Send + Sync>;

/// Type alias for transition failed callback
type TransitionFailedCallback<S, E> = Box<dyn Fn(&S, &E, &StateMachineError) + Send + Sync>;

/// Represents a state machine instance
pub struct StateMachine<S: State, E: Event> {
    /// Current state
    current_state: S,
    /// List of all defined transitions
    transitions: Vec<Transition<S, E>>,
    /// Index of transitions by (state, event) for fast lookup
    transition_index: HashMap<(S, E), usize>,
    /// History manager
    history: HistoryManager<S, E>,
    /// Whether to record history
    record_history: bool,
    /// Optional callback when state changes
    on_state_change: Option<StateChangeCallback<S, E>>,
    /// Optional callback when transition fails
    on_transition_failed: Option<TransitionFailedCallback<S, E>>,
}

impl<S: State, E: Event> StateMachine<S, E> {
    /// Create a new state machine with an initial state
    pub fn new(initial_state: S) -> Self {
        info!("Creating new state machine with initial state: {:?}", initial_state);
        Self {
            current_state: initial_state,
            transitions: Vec::new(),
            transition_index: HashMap::new(),
            history: HistoryManager::default(),
            record_history: true,
            on_state_change: None,
            on_transition_failed: None,
        }
    }

    /// Create a new state machine with history recording disabled
    pub fn new_without_history(initial_state: S) -> Self {
        let mut sm = Self::new(initial_state);
        sm.record_history = false;
        sm
    }

    /// Set the maximum number of history entries
    pub fn with_history_capacity(mut self, capacity: usize) -> Self {
        self.history = HistoryManager::new(capacity);
        self
    }

    /// Set a callback for state changes
    pub fn on_state_change(mut self, callback: impl Fn(&S, &S, &E) + Send + Sync + 'static) -> Self {
        self.on_state_change = Some(Box::new(callback));
        self
    }

    /// Set a callback for transition failures
    pub fn on_transition_failed(mut self, callback: impl Fn(&S, &E, &StateMachineError) + Send + Sync + 'static) -> Self {
        self.on_transition_failed = Some(Box::new(callback));
        self
    }

    /// Add a transition to the state machine
    pub fn add_transition(&mut self, transition: Transition<S, E>) {
        let key = (transition.from.clone(), transition.event.clone());
        let index = self.transitions.len();

        debug!(
            "Adding transition: {:?} --({:?})--> {:?}",
            transition.from, transition.event, transition.to
        );

        self.transition_index.insert(key, index);
        self.transitions.push(transition);
    }

    /// Add multiple transitions at once
    pub fn add_transitions(&mut self, transitions: Vec<Transition<S, E>>) {
        for transition in transitions {
            self.add_transition(transition);
        }
    }

    /// Get the current state
    pub fn current_state(&self) -> &S {
        &self.current_state
    }

    /// Set the current state directly (bypasses transitions)
    pub fn set_state(&mut self, state: S) {
        info!("Directly setting state from {:?} to {:?}", self.current_state, state);
        self.current_state = state;
    }

    /// Process an event and potentially transition to a new state
    pub fn process_event(&mut self, event: E) -> Result<()> {
        let start_time = Instant::now();
        debug!("Processing event: {:?}", event);

        // Find matching transition
        let transition = self.find_transition(&self.current_state, &event)?;

        let from_state = self.current_state.clone();
        let to_state = transition.to.clone();

        // Check guard condition
        if !transition.check_guard(&event) {
            let error = StateMachineError::guard_rejected(&from_state, &to_state);
            warn!("Guard rejected transition: {:?}", error);

            if self.record_history {
                self.history.push(HistoryEntry::failure(
                    from_state.clone(),
                    to_state.clone(),
                    event.clone(),
                    error.to_string(),
                ));
            }

            if let Some(callback) = &self.on_transition_failed {
                callback(&from_state, &event, &error);
            }

            return Err(error);
        }

        // Execute action
        if let Err(e) = transition.execute_action(&event) {
            warn!("Action failed: {:?}", e);

            if self.record_history {
                self.history.push(HistoryEntry::failure(
                    from_state.clone(),
                    to_state.clone(),
                    event.clone(),
                    e.to_string(),
                ));
            }

            if let Some(callback) = &self.on_transition_failed {
                callback(&from_state, &event, &e);
            }

            return Err(e);
        }

        // Perform the transition
        self.current_state = to_state.clone();
        let duration = start_time.elapsed();

        info!(
            "Transition successful: {:?} --({:?})--> {:?} ({:?})",
            from_state, event, to_state, duration
        );

        // Record history
        if self.record_history {
            self.history.push(
                HistoryEntry::success(from_state.clone(), to_state.clone(), event.clone())
                    .with_duration(duration),
            );
        }

        // Notify state change
        if let Some(callback) = &self.on_state_change {
            callback(&from_state, &to_state, &event);
        }

        Ok(())
    }

    /// Find a transition for the given state and event
    fn find_transition(&self, state: &S, event: &E) -> Result<&Transition<S, E>> {
        let key = (state.clone(), event.clone());

        match self.transition_index.get(&key) {
            Some(&index) => Ok(&self.transitions[index]),
            None => Err(StateMachineError::no_transition(state, event)),
        }
    }

    /// Check if a transition is possible for the given event
    pub fn can_process_event(&self, event: &E) -> bool {
        let key = (self.current_state.clone(), event.clone());
        self.transition_index.contains_key(&key)
    }

    /// Get all possible events from the current state
    pub fn possible_events(&self) -> Vec<&E> {
        self.transitions
            .iter()
            .filter(|t| t.from == self.current_state)
            .map(|t| &t.event)
            .collect()
    }

    /// Get all transitions
    pub fn transitions(&self) -> &[Transition<S, E>] {
        &self.transitions
    }

    /// Get the history manager
    pub fn history(&self) -> &HistoryManager<S, E> {
        &self.history
    }

    /// Get a mutable reference to the history manager
    pub fn history_mut(&mut self) -> &mut HistoryManager<S, E> {
        &mut self.history
    }

    /// Check if history recording is enabled
    pub fn is_recording_history(&self) -> bool {
        self.record_history
    }

    /// Enable or disable history recording
    pub fn set_record_history(&mut self, record: bool) {
        self.record_history = record;
    }

    /// Get the number of defined transitions
    pub fn transition_count(&self) -> usize {
        self.transitions.len()
    }

    /// Get a formatted string of all transitions
    pub fn format_transitions(&self) -> String {
        self.transitions
            .iter()
            .map(|t| {
                format!(
                    "{:?} --({:?})--> {:?}",
                    t.from, t.event, t.to
                )
            })
            .collect::<Vec<_>>()
            .join("\n")
    }

    /// Get a formatted string of the transition graph
    pub fn format_graph(&self) -> String {
        let mut graph = String::new();
        graph.push_str("State Machine Transition Graph\n");
        graph.push_str("==============================\n\n");

        // Group transitions by source state
        let mut by_state: HashMap<&S, Vec<&Transition<S, E>>> = HashMap::new();
        for t in &self.transitions {
            by_state.entry(&t.from).or_default().push(t);
        }

        for (state, transitions) in &by_state {
            graph.push_str(&format!("State {:?}:\n", state));
            for t in transitions {
                let guard_str = if t.guard.is_some() { " [guarded]" } else { "" };
                let action_str = if t.action.is_some() { " [action]" } else { "" };
                let desc_str = t.description
                    .as_ref()
                    .map(|d| format!(" - {}", d))
                    .unwrap_or_default();

                graph.push_str(&format!(
                    "  --({:?})--> {:?}{}{}{}\n",
                    t.event, t.to, guard_str, action_str, desc_str
                ));
            }
            graph.push('\n');
        }

        graph
    }
}

impl<S: State, E: Event> Debug for StateMachine<S, E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("StateMachine")
            .field("current_state", &self.current_state)
            .field("transition_count", &self.transitions.len())
            .field("history_entries", &self.history.len())
            .field("record_history", &self.record_history)
            .finish()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::transition::TransitionBuilder;

    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
    enum TestState {
        Idle,
        Running,
        Paused,
        Stopped,
    }

    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
    enum TestEvent {
        Start,
        Pause,
        Resume,
        Stop,
    }

    impl State for TestState {}
    impl Event for TestEvent {}

    fn create_test_sm() -> StateMachine<TestState, TestEvent> {
        let mut sm = StateMachine::new(TestState::Idle);

        sm.add_transition(Transition::new(TestState::Idle, TestState::Running, TestEvent::Start));
        sm.add_transition(Transition::new(TestState::Running, TestState::Paused, TestEvent::Pause));
        sm.add_transition(Transition::new(TestState::Paused, TestState::Running, TestEvent::Resume));
        sm.add_transition(Transition::new(TestState::Running, TestState::Stopped, TestEvent::Stop));
        sm.add_transition(Transition::new(TestState::Paused, TestState::Stopped, TestEvent::Stop));

        sm
    }

    #[test]
    fn test_new_state_machine() {
        let sm = StateMachine::<TestState, TestEvent>::new(TestState::Idle);
        assert_eq!(sm.current_state(), &TestState::Idle);
        assert_eq!(sm.transition_count(), 0);
        assert!(sm.is_recording_history());
    }

    #[test]
    fn test_new_without_history() {
        let sm = StateMachine::<TestState, TestEvent>::new_without_history(TestState::Idle);
        assert!(!sm.is_recording_history());
    }

    #[test]
    fn test_add_transition() {
        let mut sm = StateMachine::new(TestState::Idle);
        sm.add_transition(Transition::new(TestState::Idle, TestState::Running, TestEvent::Start));
        assert_eq!(sm.transition_count(), 1);
    }

    #[test]
    fn test_add_transitions() {
        let mut sm = StateMachine::new(TestState::Idle);
        sm.add_transitions(vec![
            Transition::new(TestState::Idle, TestState::Running, TestEvent::Start),
            Transition::new(TestState::Running, TestState::Paused, TestEvent::Pause),
        ]);
        assert_eq!(sm.transition_count(), 2);
    }

    #[test]
    fn test_process_event() {
        let mut sm = create_test_sm();
        assert_eq!(sm.current_state(), &TestState::Idle);

        sm.process_event(TestEvent::Start).unwrap();
        assert_eq!(sm.current_state(), &TestState::Running);

        sm.process_event(TestEvent::Pause).unwrap();
        assert_eq!(sm.current_state(), &TestState::Paused);

        sm.process_event(TestEvent::Resume).unwrap();
        assert_eq!(sm.current_state(), &TestState::Running);

        sm.process_event(TestEvent::Stop).unwrap();
        assert_eq!(sm.current_state(), &TestState::Stopped);
    }

    #[test]
    fn test_process_event_no_transition() {
        let mut sm = create_test_sm();
        let result = sm.process_event(TestEvent::Stop);
        assert!(result.is_err());
        assert_eq!(sm.current_state(), &TestState::Idle);
    }

    #[test]
    fn test_process_event_with_guard() {
        let mut sm = StateMachine::new(TestState::Idle);

        sm.add_transition(
            Transition::new(TestState::Idle, TestState::Running, TestEvent::Start)
                .with_guard(|_from, _to, event| *event == TestEvent::Start),
        );

        sm.process_event(TestEvent::Start).unwrap();
        assert_eq!(sm.current_state(), &TestState::Running);
    }

    #[test]
    fn test_process_event_guard_rejected() {
        let mut sm = StateMachine::new(TestState::Idle);

        sm.add_transition(
            Transition::new(TestState::Idle, TestState::Running, TestEvent::Start)
                .with_guard(|_from, _to, _event| false),
        );

        let result = sm.process_event(TestEvent::Start);
        assert!(result.is_err());
        assert_eq!(sm.current_state(), &TestState::Idle);
    }

    #[test]
    fn test_process_event_with_action() {
        let mut sm = StateMachine::new(TestState::Idle);

        sm.add_transition(
            Transition::new(TestState::Idle, TestState::Running, TestEvent::Start)
                .with_action(|_from, _to, _event| {
                    // Action executed
                    Ok(())
                }),
        );

        sm.process_event(TestEvent::Start).unwrap();
        assert_eq!(sm.current_state(), &TestState::Running);
    }

    #[test]
    fn test_process_event_action_failed() {
        let mut sm = StateMachine::new(TestState::Idle);

        sm.add_transition(
            Transition::new(TestState::Idle, TestState::Running, TestEvent::Start)
                .with_action(|_from, _to, _event| {
                    Err(StateMachineError::action_failed("test error"))
                }),
        );

        let result = sm.process_event(TestEvent::Start);
        assert!(result.is_err());
        assert_eq!(sm.current_state(), &TestState::Idle);
    }

    #[test]
    fn test_can_process_event() {
        let sm = create_test_sm();
        assert!(sm.can_process_event(&TestEvent::Start));
        assert!(!sm.can_process_event(&TestEvent::Pause));
        assert!(!sm.can_process_event(&TestEvent::Resume));
        assert!(!sm.can_process_event(&TestEvent::Stop));
    }

    #[test]
    fn test_possible_events() {
        let sm = create_test_sm();
        let events = sm.possible_events();
        assert_eq!(events.len(), 1);
        assert_eq!(events[0], &TestEvent::Start);
    }

    #[test]
    fn test_history_recording() {
        let mut sm = create_test_sm();

        sm.process_event(TestEvent::Start).unwrap();
        sm.process_event(TestEvent::Pause).unwrap();
        sm.process_event(TestEvent::Resume).unwrap();

        assert_eq!(sm.history().len(), 3);
    }

    #[test]
    fn test_history_disabled() {
        let mut sm = StateMachine::new_without_history(TestState::Idle);
        sm.add_transition(Transition::new(TestState::Idle, TestState::Running, TestEvent::Start));

        sm.process_event(TestEvent::Start).unwrap();
        assert_eq!(sm.history().len(), 0);
    }

    #[test]
    fn test_on_state_change_callback() {
        let mut sm = StateMachine::new(TestState::Idle);

        sm = sm.on_state_change(|_from, _to, _event| {
            // Callback would be called
        });

        sm.add_transition(Transition::new(TestState::Idle, TestState::Running, TestEvent::Start));
        sm.process_event(TestEvent::Start).unwrap();
    }

    #[test]
    fn test_set_state() {
        let mut sm: StateMachine<TestState, TestEvent> = StateMachine::new(TestState::Idle);
        assert_eq!(sm.current_state(), &TestState::Idle);

        sm.set_state(TestState::Running);
        assert_eq!(sm.current_state(), &TestState::Running);
    }

    #[test]
    fn test_format_transitions() {
        let sm = create_test_sm();
        let formatted = sm.format_transitions();
        assert!(formatted.contains("Idle"));
        assert!(formatted.contains("Running"));
        assert!(formatted.contains("Start"));
    }

    #[test]
    fn test_format_graph() {
        let sm = create_test_sm();
        let graph = sm.format_graph();
        assert!(graph.contains("State Machine Transition Graph"));
        assert!(graph.contains("Idle"));
        assert!(graph.contains("Running"));
    }

    #[test]
    fn test_debug() {
        let sm = create_test_sm();
        let debug_str = format!("{:?}", sm);
        assert!(debug_str.contains("Idle"));
        assert!(debug_str.contains("transition_count: 5"));
    }

    #[test]
    fn test_transition_builder() {
        let mut sm = StateMachine::new(TestState::Idle);

        let transition = TransitionBuilder::new()
            .from(TestState::Idle)
            .to(TestState::Running)
            .on(TestEvent::Start)
            .describe("Start the machine")
            .build()
            .unwrap();

        sm.add_transition(transition);
        sm.process_event(TestEvent::Start).unwrap();
        assert_eq!(sm.current_state(), &TestState::Running);
    }
}
